### PR TITLE
#737 Make it possible to dump benchmark diff report in JSON format

### PR DESCRIPTION
**Pull request by JBakamovic**
This PR brings proposed solution for functionality described in #737.

Proposed format of JSON dump is as follows:
```
[
    {
	    "measurements": [ // here comes all the measurements we've got for given benchmark
	        {
                    "real_time_other": 1.7604541402884333,
	            "cpu_time": 1.7641470201511333,
		    "time": -0.0175007535861843,
		    "real_time": 1.7918122041459086,
       		    "cpu_time_other": 1.743907157894737,
		    "cpu": -0.011472888611439235
	        },
	        ... 
	        {
		    "real_time_other": 1.8875298447986333,
      		    "cpu_time": 1.7457787153652398,
		    "time": 0.06659085312291296,
		    "real_time": 1.7696850102099235,
		    "cpu_time_other": 1.8679594937343365,
		    "cpu": 0.06998640623450084
	        }
            ],
	    "utest": { // utest section will be empty when not applicable (utest not requested)
		    "cpu_pvalue": 0.4033953048926283,
		    "time_pvalue": 0.2100750407866585,
		    "stable": false
	    },
	    "name":  "benchmark test case name",
	    "time_unit": "ms"
    },
    ... // Same for other benchmarks ...
]
```

### CODE REVIEW

File: tools/gbench/report.py [hunk 1]
```diff
@@ -345,6 +415,89 @@ def test_basic(self):
             self.assertEqual(len(parts), 7)
             self.assertEqual(expect_lines[i], parts)
 
+    def test_json_difference_report_output(self):
+        expected_output = [
+            {
+                'name': 'BM_SameTimes',
+                'measurements': [{'time': 0.0000, 'cpu': 0.0000, 'real_time': 10, 'real_time_other': 10, 'cpu_time': 10, 'cpu_time_other': 10}],
```
Comment (LebedevRI) [inline tools/gbench/report.py]: Please format the changes.

Comment (JBakamovic) [reply tools/gbench/report.py]: Will do so but I'm not sure what you're thinking on particularly? I tried to format these as much as I could so that the unit tests do not become too much spread out because of the test data. Can you please give an example of how you would like it to be?

Comment (LebedevRI) [reply tools/gbench/report.py]:
```
$ autopep8 -ia tools/compare.py tools/gbench/report.py tools/gbench/util.py
```

File: tools/gbench/report.py [hunk 2]
```diff
@@ -266,11 +274,75 @@ def get_color(res):
     return output_strs
 
 
-###############################################################################
-# Unit tests
+def get_json_difference_report(
```
Comment (LebedevRI) [inline tools/gbench/report.py]: This is bad.
Now there are two different functions that *should* only differ with the output format.
Meaning, every change will need to happen in two places.
And extra attention will need to be paid to prevent them from diverging.

Instead, `get_json_difference_report()` should be the only one,
and `generate_difference_report()` should be fed the output of `get_json_difference_report()`,
and it should just pretty-print it.

File: tools/compare.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env python
 
+import unittest
```
Comment (LebedevRI) [inline tools/compare.py]: Can you please run the autopep on git master and submit it as a pr, so this pr will be formatting-change-only clean?

File: tools/gbench/report.py [hunk 4]
```diff
@@ -34,6 +36,7 @@ def __format__(self, format):
 BC_BOLD = BenchmarkColor('BOLD', '\033[1m')
 BC_UNDERLINE = BenchmarkColor('UNDERLINE', '\033[4m')
 
+UTEST_DEFAULT_ALPHA = 0.05
```
Comment (LebedevRI) [inline tools/gbench/report.py]: I agree that the old hardcoding of `0.05` is wrong, but this is not correct either.
The specified `utest_alpha` (passed to `generate_difference_report()`) should be propagated/used.

It would be really great if you could submit this as a preparatory PR, to unclutter this PR :)

File: tools/gbench/report.py (hunk 1 above)
Comment (LebedevRI) [reply tools/gbench/report.py]: Hmm, weird, that didn't linewrap (80 col) these lines.

File: tools/gbench/report.py [hunk 5]
```diff
@@ -143,32 +146,43 @@ def extract_field(partition, field_name):
...
+    return (min_rep_cnt >= UTEST_OPTIMAL_REPETITIONS), cpu_pvalue, time_pvalue
+
+
+def print_utest(partition, utest_alpha, first_col_width, use_color=True):
+    def get_utest_color(pval):
+        return BC_FAIL if pval >= utest_alpha else BC_OKGREEN
+
+    timings_time = extract_field(partition, 'real_time')
+    timings_cpu = extract_field(partition, 'cpu_time')
+    stable_results, cpu_pvalue, time_pvalue = calc_utest(
```
Comment (LebedevRI) [inline tools/gbench/report.py]: What does `stable_results` mean here? Is this about the pvalue? Or repetition count?
Please s/`stable_results`/`sufficient_repetitions`/

File: tools/gbench/report.py [hunk 6]
```diff
@@ -143,32 +146,43 @@ def extract_field(partition, field_name):
     return [lhs, rhs]
 
 
-def print_utest(partition, utest_alpha, first_col_width, use_color=True):
-    timings_time = extract_field(partition, 'real_time')
-    timings_cpu = extract_field(partition, 'cpu_time')
-
+def calc_utest(timings_cpu, timings_time):
```
Comment (LebedevRI) [inline tools/gbench/report.py]: I think these `print_utest()` changes can count as a separate preparatory NFC refactoring,
could you please split this into separate PR?

File: tools/compare.py (hunk 3 above)
Comment (JBakamovic) [reply tools/compare.py]: Done. #739

File: tools/gbench/report.py (hunk 5 above)
Comment (JBakamovic) [reply tools/gbench/report.py]: I had it named `optimal_repetitions` previously but changed it to `stable_results` because I thought that communicating that result**s** (both pvalues) not being stable was sufficient (generic enough) to indicate the erroneous state. That way we are avoiding leaking details of how this error condition is actually evaluated by the underlying function call: `calc_utest`.

Let me know what you think about that.

File: tools/gbench/report.py (hunk 6 above)
Comment (JBakamovic) [reply tools/gbench/report.py]: Done. #740

File: tools/gbench/report.py (hunk 4 above)
Comment (JBakamovic) [reply tools/gbench/report.py]: > I agree that the old hardcoding of 0.05 is wrong, but this is not correct either.
> The specified utest_alpha (passed to generate_difference_report()) should be propagated/used.

You're actually referring to [this line](https://github.com/google/benchmark/blob/daa23894297781fdb1192a35d60eba1b5ad728d4/tools/gbench/report.py#L243). If so, I will make a change and send it another PR if that's what you would like.

File: tools/gbench/report.py [hunk 7]
```diff
@@ -232,7 +240,7 @@ def find_test(name):
             fmt_str = "{}{:<{}s}{endc}{}{:+16.4f}{endc}{}{:+16.4f}{endc}{:14.0f}{:14.0f}{endc}{:14.0f}{:14.0f}"
 
             def get_color(res):
-                if res > 0.05:
+                if res > UTEST_DEFAULT_ALPHA:
```
Comment (LebedevRI) [inline tools/gbench/report.py]: Magic numbers are fun, aren't they? :)
Although this is also `0.05`, this is *not* `UTEST_DEFAULT_ALPHA`,
these are used to color the change of the timers, note the `elif`.

File: tools/gbench/report.py [hunk 8]
```diff
@@ -192,21 +206,15 @@ def generate_difference_report(
         json2,
         display_aggregates_only=False,
         utest=False,
-        utest_alpha=0.05,
+        utest_alpha=UTEST_DEFAULT_ALPHA,
```
Comment (LebedevRI) [inline tools/gbench/report.py]: Ok, i think this can be changed.

File: tools/gbench/report.py [hunk 9]
```diff
@@ -437,7 +632,7 @@ def test_utest(self):
         ]
         json1, json2 = self.load_results()
         output_lines_with_header = generate_difference_report(
-            json1, json2, utest=True, utest_alpha=0.05, use_color=False)
+            json1, json2, utest=True, utest_alpha=UTEST_DEFAULT_ALPHA, use_color=False)
```
Comment (LebedevRI) [inline tools/gbench/report.py]: Ah wait, I'm not sure it is a good idea to change this.
This is a test, and it is very specifically testing what happens with `utest_alpha=0.05`.
If you pass `UTEST_DEFAULT_ALPHA`, and then change it, the test will need updating.

File: tools/gbench/report.py [hunk 10]
```diff
@@ -505,7 +778,7 @@ def test_utest(self):
         json1, json2 = self.load_results()
         output_lines_with_header = generate_difference_report(
             json1, json2, display_aggregates_only=True,
-            utest=True, utest_alpha=0.05, use_color=False)
+            utest=True, utest_alpha=UTEST_DEFAULT_ALPHA, use_color=False)
```
Comment (LebedevRI) [inline tools/gbench/report.py]: Same.

File: tools/gbench/report.py (hunk 7 above)
Comment (JBakamovic) [reply tools/gbench/report.py]: Ok, I misunderstood this one then.

File: tools/gbench/report.py (hunk 9 above)
Comment (JBakamovic) [reply tools/gbench/report.py]: Alright, it makes sense. Then I think reverting the whole change makes most sense because most occurrences of it are in unit tests. Do you agree?

Comment (LebedevRI) [reply tools/gbench/report.py]: I have re-read the diff, and i'm not sure what is cleaner.
Out of all the places where `0.05` is used, only one can be replaced with `UTEST_DEFAULT_ALPHA`.
So yes, the situation will not be great either way. I'd undo these `UTEST_DEFAULT_ALPHA` changes. :/

Comment (JBakamovic) [reply tools/gbench/report.py]: Agreed. I will revert the change.

File: tools/gbench/report.py (hunk 2 above)
Comment (JBakamovic) [reply tools/gbench/report.py]: Alright, I think I'm on my way finishing this change. Please see 9021718. `generate_difference_report` is now a consumer of data fed by `get_json_difference_report`.

One more thing I've yet to be done is to re-implement the unit-tests to complement this change.

Comment (JBakamovic) [reply tools/gbench/report.py]: One more thing I would like to note and which is related to your comment in #740 is that `print_utest` now became much less involved and looks more natural to what is supposed to actually do.

File: tools/gbench/report.py [hunk 11]
```diff
@@ -215,64 +225,112 @@ def find_test(name):
...
-
-    return output_strs
-
+            have_optimal_repetitions, cpu_pvalue, time_pvalue = calc_utest(
+                extract_field(partition, 'real_time'),
+                extract_field(partition, 'cpu_time')
+            )
+            if cpu_pvalue and time_pvalue:
+                utest_results = {
+                    'stable': have_optimal_repetitions,
```
Comment (LebedevRI) [inline tools/gbench/report.py]: I'm not sure `stable` is correct word here.
It should be `sufficient_observation_count`.
It *looks* like they say the same thing, but the meaning is slightly different.

Comment (JBakamovic) [reply tools/gbench/report.py]: `sufficient_observation_count` seems a bit overly wordy IMO. Perhaps, `optimal_repetitions` if `stable` is not satisfying?

Newest changes are up so please have a look at them when you find it appropriate.

Comment (LebedevRI) [reply tools/gbench/report.py]: `sufficient` has a very direct meaning based on https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html
> Use only when the number of observation in each sample is > 20

What is the definition of `optimal`, `stable` in this case?
We don't know that having less/more repetitions would have degraded the u test meaningfulness..

Comment (JBakamovic) [reply tools/gbench/report.py]: > What is the definition of optimal, stable in this case?

I was following the wording of existing code (e.g. `UTEST_OPTIMAL_REPETITIONS`) so if we decide that `optimal` is not good enough changes should be done to that parts of the source code as well.

> sufficient has a very direct meaning based on https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html

Alright, makes sense. So `sufficient_observation_count` it is?

Comment (JBakamovic) [reply tools/gbench/report.py]: @LebedevRI ping :)

**Comment by fsaintjacques** [comment]
ping?

**Comment by JBakamovic** [comment]
> ping?

@fsaintjacques I believe there is not enough interest from maintainers side and PR is now somewhat stale.

**Comment by dominichamon** [comment]
my apologies, i missed that this was ready for further reviews. If you'd be willing to resolve the conflicts or restart the PR i'd be happy to take a look.

**Comment by JBakamovic** [comment]
@dominichamon Give me few days and I'll see what I can do. It's been a while.

**Comment by JBakamovic** [comment]
I am closing this PR in favor of #1042 .
